modules/__tests__/describeQueries.js

```diff
@@ -45,6 +45,48 @@ function describeQueries(createHistory) {
       })
     })
 
+    describe('in push', function () {
+      it('works', function (done) {
+        let steps = [
+          function (location) {
+            expect(location.pathname).toEqual('/')
+            expect(location.search).toEqual('')
+            expect(location.query).toEqual({})
+            expect(location.state).toEqual(null)
+            expect(location.action).toEqual(POP)
+
+            history.push({
+              pathname: '/home',
+              query: { the: 'query value' },
+              state: { the: 'state' }
+            })
+          },
+          function (location) {
+            expect(location.pathname).toEqual('/home')
+            expect(location.search).toEqual('?the=query+value')
+            expect(location.query).toEqual({ the: 'query value' })
+            expect(location.state).toEqual({ the: 'state' })
+            expect(location.action).toEqual(PUSH)
+
+            history.push({
+              ...location,
+              query: { other: 'query value' },
+              state: { other: 'state' }
+            })
+          },
+          function (location) {
+            expect(location.pathname).toEqual('/home')
+            expect(location.search).toEqual('?other=query+value')
+            expect(location.query).toEqual({ other: 'query value' })
```

Comment:
It's technically not a breaking change because you can't currently push an object at all, so there's no old behavior to break. The existing behavior covered by e.g. this test case: https://github.com/taion/history/blob/push-replace-location-object/modules/__tests__/describeQueries.js#L178-L184 still works.

I think for a 2.x release, we should always overwrite the query and change that linked test case above. I dunno. The workaround I have right now is quite ugly, but I'd like to minimize the number of major version bumps.